### PR TITLE
Compatiblity with GraphQL 3.0.0

### PR DESCRIPTION
**Pull request by BoBiene**
Based on graphql-dotnet#194

Currently the Unit-Test Tests.Adapters.Engine.Bugs.BugUnexpectedTypeTests.Can_Resolve_Null_NonNull_Query fails.
@clangelov can you please assist fixing this?

### CODE REVIEW

**Comment by clangelov** [comment]
> Based on #194
> 
> Currently the Unit-Test Tests.Adapters.Engine.Bugs.BugUnexpectedTypeTests.Can_Resolve_Null_NonNull_Query fails.
> @clangelov can you please assist fixing this?

Sure :) I think the issue is related to your changes in the GraphTypeAdapter.cs,  you need to use the same method overload as from my PR  https://github.com/graphql-dotnet/conventions/pull/205 for Schema.RegisterTypes(...)  and probably it will resolve the issue. I will be able to give you more feedback on Thursday

**Comment by clangelov** [comment]
> Based on #194
> 
> Currently the Unit-Test Tests.Adapters.Engine.Bugs.BugUnexpectedTypeTests.Can_Resolve_Null_NonNull_Query fails.
> @clangelov can you please assist fixing this?

In the GraphTypeAdapter DeriveSchema method you need to change it to 

```
// build params IGraphType[] types istead of Types[]
            var possibleTypes = interfaces 
                .Where(t => !t.IsIgnored)
                .SelectMany(t => t.PossibleTypes)
                .Select(typeInfo => (typeInfo.IsIgnored || !typeInfo.IsNullable || typeInfo.Interfaces.Any(x => x.IsIgnored == true))
                    ? null
                    : DeriveType(typeInfo)
                )
                .Where(x => x != null)
                .GroupBy(t => t.Name)
                .Select(g => g.First())
                .ToArray();
            var schema = new Schema(new FuncDependencyResolver(DeriveTypeFromTypeInfo)) // this one can remain the same 
            {
                Query = DeriveOperationType(schemaInfo.Query),
                Mutation = DeriveOperationType(schemaInfo.Mutation),
                Subscription = DeriveOperationType(schemaInfo.Subscription),
            };
            schema.RegisterTypes(possibleTypes); // register proper IGraphType[] types which won't fail with unrecognized types
```

and the test should pass. It's kind-a difficult to explain why this is needed, but when you have larger and more complex schema, when you pass Type[] it just fails to properly register the graph types and queries start to fail

**Comment by BoBiene** [comment]
Thank you @clangelov 👍
@tlil can you please review this Pullrequest

**Comment by BoBiene** [comment]
@sungam3r, can you please take a look at this?

**Comment by tlil** [comment]
Yes, will review this weekend. Thanks!

**Comment by sungam3r** [comment]
I can review this next week after my vacation.

**Comment by yokan87** [comment]
I just fixed two issues in this branch I had with subscriptions (Type name generation and subscription execution).
Tests are still running successfully.

File: src/GraphQL.Conventions/Adapters/Engine/GraphQLEngine.cs [hunk 1]
```diff
@@ -261,10 +253,9 @@ public GraphQLExecutor NewExecutor(IRequestDeserializer requestDeserializer = nu
             return this;
         }
 
-        public string SerializeResult(ExecutionResult result)
-        {
-            return _documentWriter.Write(result);
-        }
+        public string SerializeResult(ExecutionResult result) => SerializeResultAsync(result).GetAwaiter().GetResult();
```
Comment (sungam3r) [inline src/GraphQL.Conventions/Adapters/Engine/GraphQLEngine.cs:256]: I strongly do not recommend you to have sync over async methods.

File: src/GraphQL.Conventions/Adapters/ResolutionContext.cs [hunk 2]
```diff
@@ -57,15 +58,15 @@ public void SetArgument(string name, object value)
...
 
-        public IDependencyInjector DependencyInjector => (FieldContext.UserContext as IDependencyInjectorAccessor)?.DependencyInjector;
+        public IDependencyInjector DependencyInjector => FieldContext.GetDependencyInjector();
 
         public GraphFieldInfo FieldInfo { get; private set; }
 
         public CancellationToken CancellationToken => FieldContext.CancellationToken;
 
-        public IEnumerable<string> Path => FieldContext.Path;
+        public IEnumerable<string> Path => FieldContext.Path.Select(o=> o?.ToString());
```
Comment (sungam3r) [inline src/GraphQL.Conventions/Adapters/ResolutionContext.cs:69]:
```suggestion
        public IEnumerable<string> Path => FieldContext.Path.Select(o => o?.ToString());
```

File: src/GraphQL.Conventions/Adapters/Types/IdGraphType.cs [hunk 3]
```diff
@@ -1,11 +1,18 @@
 using GraphQL.Conventions.Adapters.Types.Extensions;
+using GraphQL.Conventions.Types.Descriptors;
 using GraphQL.Language.AST;
+using GraphQL.Types;
 using static GraphQL.Conventions.Utilities;
 
 namespace GraphQL.Conventions.Adapters.Types
 {
-    public class IdGraphType : GraphQL.Types.IdGraphType
+    public class IdGraphType : ScalarGraphType
```
Comment (sungam3r) [inline src/GraphQL.Conventions/Adapters/Types/IdGraphType.cs:9]: Why do we need this type at all, if GraphL.NET already has one?

File: src/GraphQL.Conventions/GraphQL.Conventions.csproj [hunk 4]
```diff
@@ -26,36 +26,28 @@
...
-  <PropertyGroup>
-    <GenerateDocumentationFile>false</GenerateDocumentationFile>
-  </PropertyGroup>
+	<PropertyGroup>
+		<LangVersion>7.3</LangVersion>
+	</PropertyGroup>
+
+	<ItemGroup>
+		<PackageReference Include="GraphQL" Version="3.0.0" />
+		<PackageReference Include="GraphQL.NewtonsoftJson" Version="3.0.0" />
```
Comment (sungam3r) [inline src/GraphQL.Conventions/GraphQL.Conventions.csproj:35]: I recommend that you follow a general strategy and not make dependencies on a particular serializer.

File: src/GraphQL.Conventions/GraphQL.Conventions.csproj [hunk 5]
```diff
@@ -26,36 +26,28 @@
...
+
+	<PropertyGroup Condition=" '$(Configuration)' == 'Debug' ">
+		<DefineConstants>$(DefineConstants);DEBUG</DefineConstants>
+	</PropertyGroup>
+	<PropertyGroup Condition=" '$(Configuration)' == 'Release' ">
+		<DefineConstants>$(DefineConstants);RELEASE</DefineConstants>
+	</PropertyGroup>
+
+	<PropertyGroup>
+		<GenerateDocumentationFile>false</GenerateDocumentationFile>
```
Comment (sungam3r) [inline src/GraphQL.Conventions/GraphQL.Conventions.csproj:50]: Why?

File: src/GraphQL.Conventions/GraphQL.Conventions.csproj [hunk 6]
```diff
@@ -26,36 +26,28 @@
...
+		<PackageReference Include="System.Reflection.Extensions" Version="4.3.0" />
+		<PackageReference Include="System.Reflection.TypeExtensions" Version="4.6.0" />
+	</ItemGroup>
+
+	<PropertyGroup Condition=" '$(Configuration)' == 'Debug' ">
+		<DefineConstants>$(DefineConstants);DEBUG</DefineConstants>
+	</PropertyGroup>
+	<PropertyGroup Condition=" '$(Configuration)' == 'Release' ">
+		<DefineConstants>$(DefineConstants);RELEASE</DefineConstants>
+	</PropertyGroup>
```
Comment (sungam3r) [inline src/GraphQL.Conventions/GraphQL.Conventions.csproj:47]:
```suggestion
```
I'm sure this already works by default.

File: src/GraphQL.Conventions/GraphQL.Conventions.csproj [hunk 7]
```diff
@@ -26,36 +26,28 @@
...
-  </PropertyGroup>
-  <PropertyGroup Condition=" '$(Configuration)' == 'Release' ">
-    <DefineConstants>$(DefineConstants);RELEASE</DefineConstants>
-  </PropertyGroup>
-
-  <PropertyGroup>
-    <GenerateDocumentationFile>false</GenerateDocumentationFile>
-  </PropertyGroup>
+	<PropertyGroup>
+		<LangVersion>7.3</LangVersion>
```
Comment (sungam3r) [inline src/GraphQL.Conventions/GraphQL.Conventions.csproj:30]:
```suggestion
		<LangVersion>latest</LangVersion>
```

File: src/GraphQL.Conventions/Web/RequestHandler.cs [hunk 8]
```diff
@@ -275,14 +275,18 @@ public async Task<Response> ProcessRequest(Request request, IUserContext userCon
...
             public Response Validate(Request request)
             {
-                var result = _engine.Validate(request.QueryString);
-                return new Response(request, result);
+                return ValidateAsync(request).Result;
+            }
+            public async Task<Response> ValidateAsync(Request request)
+            {
+                var result = await _engine.Validate(request.QueryString);
+                return  new Response(request, result);
```
Comment (sungam3r) [inline src/GraphQL.Conventions/Web/RequestHandler.cs:289]:
```suggestion
                return new Response(request, result);
```

File: src/GraphQL.Conventions/Execution/UserContextExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+using GraphQL.Execution;
+using GraphQL.Types;
+
+namespace GraphQL.Conventions.Execution
+{
+    public static class UserContextExtensions
+    {
+        public static IUserContext GetUserContext(this ExecutionContext context) => GetValue<IUserContext>(context.UserContext);
+        public static IUserContext GetUserContext(this ResolveFieldContext context) => GetValue<IUserContext>(context.UserContext);
+        public static IUserContext GetUserContext<T>(this ResolveFieldContext<T> context) => GetValue<IUserContext>(context.UserContext);
```
Comment (sungam3r) [inline src/GraphQL.Conventions/Execution/UserContextExtensions.cs:11]: Maybe it worth changing ResolveFieldContext to IResolveFieldContext everywhere in this file?

File: test/GraphQL.Conventions.Tests/Adapters/Engine/ErrorTests.cs [hunk 10]
```diff
@@ -9,6 +9,8 @@ namespace GraphQL.Conventions.Tests.Adapters.Engine
 {
     public class ErrorTests : TestBase
     {
+        private const bool CanExposeExceptionData = false; // is currently not possible.
```
Comment (sungam3r) [inline test/GraphQL.Conventions.Tests/Adapters/Engine/ErrorTests.cs:12]: It is possible now, see `IErrorInfoProvider`.

File: test/GraphQL.Conventions.Tests/GraphQL.Conventions.Tests.csproj [hunk 11]
```diff
@@ -1,4 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk" ToolsVersion="15.0">
+﻿<Project Sdk="Microsoft.NET.Sdk" ToolsVersion="15.0">
```
Comment (sungam3r) [inline test/GraphQL.Conventions.Tests/GraphQL.Conventions.Tests.csproj:1]: Consider changing test TFM to netcoreapp3.1.

**Review by sungam3r — CHANGES_REQUESTED** [review]
I didn't look in great detail. I do not use this project and therefore am not an expert. I have seen two major flaws:
1. Sync over async. I propose remove such methods and keep only async ones.
2. Dependency on GraphQL serializer. I propose to remove dependency and allow to user decide which to use.

**Comment by sungam3r** [comment]
@BoBiene reviewed

**Comment by tlil** [comment]
Seems like this has now been incorporated into https://github.com/graphql-dotnet/conventions/pull/194#issuecomment-719232580? Does that mean we can close this as a dup, @BoBiene and @schabannes?

**Pull request closed without merging**